### PR TITLE
Chapter 7: Sign up a user

### PR DESCRIPTION
**Pull request by JOlsen-Neslo**

### CODE REVIEW

File: app/helpers/users_helper.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module UsersHelper
+
```
Comment (bradleymarques) [inline app/helpers/users_helper.rb]:
```suggestion
```

File: app/helpers/users_helper.rb [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+  def gravatar_for(user)
+    gravatar_id = Digest::MD5::hexdigest(user.email.downcase)
+    gravatar_url = "https://secure.gravatar.com/avatar/#{gravatar_id}"
+    image_tag(gravatar_url, alt: user.name, class: "gravatar")
+  end
+
```
Comment (bradleymarques) [inline app/helpers/users_helper.rb]:
```suggestion
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -1,4 +1,26 @@
...
+    else
+      render 'new'
+    end
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit(:name, :email, :password,
+                                 :password_confirmation)
```
Comment (bradleymarques) [inline app/controllers/users_controller.rb]: Please check your indentation settings in RubyMine.  We'd prefer this style:

```ruby
params.require(:user).permit(
  :name,
  :email,
  :password,
  :password_confirmation
)
```

Or
```ruby
params
  .require(:user)
  .permit(
    :name,
    :email,
    :password,
    :password_confirmation
  )
```

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -1,4 +1,26 @@
...
+  end
+
   def new
+    @user = User.new
+  end
+
+  def create
+    @user = User.new(user_params) # Not the final implementation!
+    if @user.save
+      flash[:success] = "Welcome to the Sample App!"
```
Comment (bradleymarques) [inline app/controllers/users_controller.rb]: Would be better to use I18n to localize this string ;)

File: test/integration/users_signup_test.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
+require 'test_helper'
+
+class UsersSignupTest < ActionDispatch::IntegrationTest
+  test "valid signup information" do
+    get signup_path
+    assert_difference 'User.count', 1 do
+      post_via_redirect users_path, params: {user: {name: "Example User",
+                                                    email: "user@example.com",
+                                                    password: "password",
+                                                    password_confirmation: "password"}}
```
Comment (bradleymarques) [inline test/integration/users_signup_test.rb]: Also check your settings for spaces after hashes.  `{ this: good }` `{this: bad}`

Comment (bradleymarques) [reply test/integration/users_signup_test.rb]: The previous comment around multi-line hashes applies here too (see https://github.com/airbnb/ruby#multiline-hashes)

File: test/integration/users_signup_test.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+    assert_template 'users/show'
+  end
+
+  test "invalid signup information" do
+    get signup_path
+    assert_no_difference 'User.count' do
+      post users_path, params: {user: {name: "",
+                                       email: "user@invalid",
+                                       password: "foo",
+                                       password_confirmation: "bar"}}
```
Comment (bradleymarques) [inline test/integration/users_signup_test.rb]: https://github.com/airbnb/ruby#multiline-hashes

File: app/views/shared/_error_messages.html.erb [hunk 7]
```diff
@@ -0,0 +1,12 @@
+<% if @user.errors.any? %>
+    <div id="error_explanation">
+        <div class="alert alert-danger">
+            The form contains <%= pluralize(@user.errors.count, "error") %>.
+        </div>
+        <ul>
+            <% @user.errors.full_messages.each do |msg| %>
```
Comment (c13e) [inline app/views/shared/_error_messages.html.erb]: One fun & instructive thing to do is to explore the structure of `ActiveModel::Errors` in a Rails console (having `pry` installed):
```ruby
(main)> user = User.new
=> #<User:0x00007fe6ddb0de40
 id: nil,
 ....

(main)> user.errors.count
=> 0

# errors only get populated after running validations
(main)> user.valid?
=> false

(main)> cd user.errors
(#<ActiveModel::Errors:0x00007fe6df8786b0>)> ls
Enumerable#methods: 
  all?            each_cons         first     max        reduce        take                                  
  any?            each_entry        flat_map  max_by     reject        take_while                            
  chunk           each_slice        grep      member?    reverse_each  to_h                                  
  chunk_while     each_with_index   grep_v    min        select        to_json                               
  collect         each_with_object  group_by  min_by     slice_after   to_json_with_active_support_encoder   
  collect_concat  entries           index_by  minmax     slice_before  to_json_without_active_support_encoder
  cycle           exclude?          inject    minmax_by  slice_when    to_set                                
  detect          find              lazy      none?      sort          uniq                                  
  drop            find_all          many?     one?       sort_by       zip                                   
  drop_while      find_index        map       partition  sum         
ActiveModel::Errors#methods: 
  []   add_on_blank  as_json  count   empty?         full_messages_for  has_key?  keys      size     to_xml
  []=  add_on_empty  blank?   delete  full_message   generate_message   include?  messages  to_a     values
  add  added?        clear    each    full_messages  get                key?      set       to_hash
self.methods: __pry__
instance variables: @base  @messages
locals: _  __  _dir_  _ex_  _file_  _in_  _out_  _pry_

(#<ActiveModel::Errors:0x00007fe6df8786b0>)> messages
=> {:password=>["can't be blank", "is too short (minimum is 8 characters)", "is invalid"], :username=>["can't be blank"]}

# feel free to explore more stuff
```

One of the most valuable skills you can learn as a Ruby/Rails developer is getting familiar with using `pry`commands to quickly... `pry` out information from data. It's probably my favourite Ruby gem :) Here's an article outlining some of the cool things you can do with it https://medium.com/rubyinside/powering-your-ruby-rails-development-with-pry-3d5dbd2a8b80

Here's a cheatsheet I've also put together for useful things one can do with pry [pry.zip](https://github.com/JOlsen-Neslo/rails-app/files/3690858/pry.zip)

Comment (ps-taylor) [reply app/views/shared/_error_messages.html.erb]: awesome comment @c13e

Comment (JOlsen-Neslo) [reply app/views/shared/_error_messages.html.erb]: @c13e thanks! This is awesome

**Pull request merged**